cpp/src/gandiva/llvm_external_ir_store.h

```diff
@@ -0,0 +1,40 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <arrow/status.h>
+#include <gandiva/visibility.h>
+#include <llvm/Support/MemoryBuffer.h>
+#include <memory>
+#include <vector>
+
+namespace gandiva {
+using arrow::Status;
+
+class GANDIVA_EXPORT LLVMExternalIRStore {
+ public:
+  /// \brief add an LLVM IR to the store from a given bitcode file path
+  static Status Add(const std::string& bitcode_file_path);
```

Comment:
Why are all these methods static? It would be much more flexible if this was a proper class that one can instantiate. Is there a technical constraint that mandates this?

Reply:
> It would be much more flexible if this was a proper class that one can instantiate

Currently, users of Gandiva uses `Projector`/`Filter` API to work with the expression, and projector/filter and their internal llvm generator/engine are created every time a new projector/filter is created, and function registry is used internally by the engine. However, we don't want users to register their functions every time, so there needs to have a global state storing the functions users register up front. Initially I implemented this a singleton to store the global state, and after discussing with @js8544 (see above), I change the singleton to static methods + an internal static variable to store the global state. Please advise if there is other approach for addressing this concern. Thanks.

Reply:
I think the way this should be done is to have a `LLVMExternalIRStore` class with non-static method and then a global singleton that can be retrieved through a specific function:
```c++
std::shared_ptr<LLVMExternalIRStore> default_external_ir_store();
```

Then the `LLVMGenerator` created by projectors and filters can use the `default_external_ir_store()` internally.

`Projector` and `Filter` could also optionally take a `LLVMExternalIRStore*` argument.

Reply:
Gandiva has many function related registries such as FunctionRegistry, ExportedFuncsRegistry, FunctionHolderRegistry. They are all global instances/singletons. I don't find it necessary to make LLVMExternalIRStore special.

Reply:
Well, fair enough, but it seems weird to want to perpetuate an inflexible design.

Reply:
Yeah I do agree that the current design is not ideal. Maybe we can have a separate PR to change them altogether.

Reply:
> I think the way this should be done is to have a LLVMExternalIRStore class with non-static methods and then a global singleton that can be retrieved through a specific function
> std::shared_ptr<LLVMExternalIRStore> default_external_ir_store();

> Projector and Filter could also optionally take a LLVMExternalIRStore* argument.

> > Maybe we can have a separate PR to change them altogether

You can check out the API for Gandiva users for function registration below, if we expose such API ((https://github.com/apache/arrow/pull/38116#issuecomment-1771223510)) for function registration, the `LLVMExternalIRStore` will be an internal API and if we agree and decide to use singleton pattern for retrieving/managing these global states in Gandiva, I may help to do some refactoring later (in a separated PR) to address the concern here. What do you think?

Reply:
See the changes here:
https://github.com/apache/arrow/pull/38116#issuecomment-1772407022